File: app/GOExercise/GOExercise/Services/CatStateResolver.swift
import Foundation

enum CatStateResolver {
    static func resolve(
        todayStatus: DailyStatus,
        now: Date,
        yesterdayAchieved: Bool,
        streakExtendedThisRun: Bool,
        calendar: Calendar = .current
    ) -> CatState {
        if streakExtendedThisRun {
            return .streakExtended
        }

        if todayStatus == .todayAchieved || todayStatus == .achieved {
            return .celebrating
        }

        if todayStatus == .rest {
            return .resting
        }

        if todayStatus == .todayPending, !yesterdayAchieved {
            return .encouraging
        }

        let hour = calendar.component(.hour, from: now)
        switch hour {
        case ..<12:
            return .waitingMorning
        case 12..<18:
            return .worriedNoon
        default:
            return .beggingNight
        }
    }
}

File: app-android/app/src/main/java/com/goexercise/app/domain/CatStateResolver.kt
package com.goexercise.app.domain

import java.time.LocalDateTime

/**
 * 今日の状態・時刻から猫の表示状態を決める。iOS `CatStateResolver.swift` の 1:1 移植。
 * 優先順位: streakExtended > 達成(celebrating) > 休息(resting) > 昨日未達成の復帰応援(encouraging)
 * > 時間帯(朝待機/昼心配/夜おねだり)。`now` は時刻判定にだけ使うので LocalDateTime で受ける。
 */
object CatStateResolver {
    fun resolve(
        todayStatus: DailyStatus,
        now: LocalDateTime,
        yesterdayAchieved: Boolean,
        streakExtendedThisRun: Boolean,
    ): CatState {
        if (streakExtendedThisRun) return CatState.StreakExtended

        if (todayStatus == DailyStatus.TodayAchieved || todayStatus == DailyStatus.Achieved) {
            return CatState.Celebrating
        }

        if (todayStatus == DailyStatus.Rest) return CatState.Resting

        if (todayStatus == DailyStatus.TodayPending && !yesterdayAchieved) {
            return CatState.Encouraging
        }

        return when (now.hour) {
            in 0 until 12 -> CatState.WaitingMorning
            in 12 until 18 -> CatState.WorriedNoon
            else -> CatState.BeggingNight
        }
    }
}
